### PR TITLE
exercise_3

### PR DESCRIPTION
**Pull request by anetso18**

### CODE REVIEW

File: exercise_3/IPerson.cpp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#include "IPerson.h"
+#include <string>
+#include <memory>
+
+std::string IPerson::getName() const
```
Comment (mjtieto) [inline exercise_3/IPerson.cpp]: The implementation in the interface is not needed. `IPerson` is meant to be a simple interface and not implement anything

File: exercise_3/IPerson.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+#include <string>
+
+class IPerson
+{
+public:
+	virtual std::string getName() const = 0;
+	virtual ~IPerson() = default;
+
+private:
+	std::string name;
```
Comment (mjtieto) [inline exercise_3/IPerson.h]: The member is not needed. It should be stored in the derived classes

File: exercise_3/IRecord.h [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+public:
+
+	virtual int getId() const = 0;
+	virtual std::string getFormatted() const = 0;
+	IRecord( int id);
+	IRecord();
+	virtual ~IRecord() = default;
+
+private:
+	 int id;
```
Comment (mjtieto) [inline exercise_3/IRecord.h]: `IRecord` should be a simple interface. Members and implemented methods should be in derived classes.

File: exercise_3/Student.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#include "Student.h"
+#include <iostream>
+#include <sstream>
+#include <memory>
+
+Student::Student(int id, std::string name)
+	:id( id ), name{ std::move( name ) }
```
Comment (mjtieto) [inline exercise_3/Student.cpp]: Decide whether you prefer brackets or parens (:

File: exercise_3/main.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+}
+
+
+int main(int argc, char* argv[])
+{
+	std::vector<std::pair<int, std::shared_ptr<IRecord>>> exams;
+	std::vector<std::pair<int, std::shared_ptr<IRecord>>> courses;
+	std::vector<std::pair<int, std::shared_ptr<IRecord>>> teachers;
+	std::vector<std::pair<int, std::shared_ptr<IRecord>>> students;
```
Comment (mjtieto) [inline exercise_3/main.cpp]: Two topics:
1. Records could be `unique_ptr`s. You don't need to share their ownership.
2. Why do you store pairs? Ids are already stored in the record, so you don't need the pair. A simple vector of records would be sufficient: `std::vector<std::unique_ptr<IRecord>>`.

File: exercise_3/main.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+				break;
+			default:
+				throw std::invalid_argument("Invalid record type");
+            }
+		}
+	}
+	else
+	{
+		std::cerr << "Unable to open file";
+		exit(1);
```
Comment (mjtieto) [inline exercise_3/main.cpp]: It's better to return a result from function than call exit directly.
Additionally, please decide whether you want to throw an exception/return result/exit (which I don't recommend).

Personally, I'd create an enum and return it from the function, e.g.:
```cpp
enum class LoadingResult {
    Ok,
    FileNotFound,
    UnknownRecordType
}
```

File: exercise_3/main.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	{
+		std::string line;
+		char separator = ' ';
+		std::string substr;
+
+		while (std::getline(in, line))
+		{
+			std::replace(line.begin(), line.end(), ',', separator);
+			std::stringstream stream(line);
+			std::getline(stream, substr, separator);
```
Comment (mjtieto) [inline exercise_3/main.cpp]: You could use `operator >>`. It's more convenient.

File: exercise_3/main.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+			return std::make_shared<Teacher>(id,name);
+			break;
+		default:
+			throw std::invalid_argument("Invalid record type");
+			break;
+		}
+	}
+
+
+template<class T>
```
Comment (mjtieto) [inline exercise_3/main.cpp]: The template parameter is not needed.

File: exercise_3/main.cpp [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		}
+	}
+	else
+	{
+		std::cerr << "Unable to open file";
+		exit(1);
+	}
+}
+
+template<class T>
```
Comment (mjtieto) [inline exercise_3/main.cpp]: The template is not needed

File: exercise_3/main.cpp [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+void writeObjectsToFile( std::string fileName, std::vector<std::pair<int, std::shared_ptr<IRecord>>>& container)
+{
+	std::ofstream output(fileName);
+	if (output)
+	{
+		sort(container.begin(), container.end());
+		for (auto &record: container)
```
Comment (mjtieto) [inline exercise_3/main.cpp]: `const auto&`

**Review by mjtieto — CHANGES_REQUESTED** [review]

File: exercise_3/IPerson.cpp (hunk 1 above)
Comment (anetso18) [reply exercise_3/IPerson.cpp]: deleted

File: exercise_3/IPerson.h (hunk 2 above)
Comment (anetso18) [reply exercise_3/IPerson.h]: removed

File: exercise_3/IRecord.h (hunk 3 above)
Comment (anetso18) [reply exercise_3/IRecord.h]: removed

File: exercise_3/Student.cpp (hunk 4 above)
Comment (anetso18) [reply exercise_3/Student.cpp]: changed to {}

File: exercise_3/main.cpp (hunk 5 above)
Comment (anetso18) [reply exercise_3/main.cpp]: 1. changed to unique_ptr s
2. changed to std::vector<std::unique_ptr<IRecord>>

File: exercise_3/main.cpp (hunk 8 above)
Comment (anetso18) [reply exercise_3/main.cpp]: removed

File: exercise_3/main.cpp (hunk 9 above)
Comment (anetso18) [reply exercise_3/main.cpp]: removed

File: exercise_3/main.cpp (hunk 10 above)
Comment (anetso18) [reply exercise_3/main.cpp]: changed to const auto&

File: exercise_3/main.cpp (hunk 7 above)
Comment (anetso18) [reply exercise_3/main.cpp]: operator >> used

File: exercise_3/main.cpp (hunk 6 above)
Comment (anetso18) [reply exercise_3/main.cpp]: proceeded with enum

File: exercise_3/IRecord.cpp [hunk 11]
```diff
@@ -1,12 +1 @@
...
-}
-
-IRecord::IRecord() {}
-
-int IRecord::getId() const {
-
-	return id;
-}
\ No newline at end of file
+#include "IRecord.h"
```
Comment (mjtieto) [inline exercise_3/IRecord.cpp]: So, the `IRecord.cpp` is not needed anymore

File: exercise_3/main.cpp [hunk 12]
```diff
@@ -69,46 +71,57 @@ void loadObjectsFromFile(const std::string& argv, std::vector<std::pair<int, std
...
 
-void writeObjectsToFile( std::string fileName, std::vector<std::pair<int, std::shared_ptr<IRecord>>>& container)
+void writeObjectsToFile( std::string fileName, std::vector< std::unique_ptr<IRecord>>& container)
 {
 	std::ofstream output(fileName);
 	if (output)
 	{
-		sort(container.begin(), container.end());
-		for (auto &record: container) 
+		sort(container.begin(), container.end(), compareRecords());
```
Comment (mjtieto) [inline exercise_3/main.cpp]: That's a perfect place for a lambda (:
```cpp
sort(container.begin(), container.end(), [](const auto& lhs, const auto& rhs) { return lhs->getId() < rhs->getId(); };
```

File: exercise_3/main.cpp [hunk 13]
```diff
@@ -118,20 +131,19 @@ void writeObjectsToFile( std::string fileName, std::vector<std::pair<int, std::s
...
-	std::vector<std::pair<int, std::shared_ptr<IRecord>>> exams;
-	std::vector<std::pair<int, std::shared_ptr<IRecord>>> courses;
-	std::vector<std::pair<int, std::shared_ptr<IRecord>>> teachers;
-	std::vector<std::pair<int, std::shared_ptr<IRecord>>> students;
+	std::vector< std::unique_ptr<IRecord>> exams;
+	std::vector< std::unique_ptr<IRecord>> courses;
+	std::vector< std::unique_ptr<IRecord>> teachers;
+	std::vector< std::unique_ptr<IRecord>> students;
 
 	loadObjectsFromFile(argv[1],exams, courses, teachers,students);
```
Comment (mjtieto) [inline exercise_3/main.cpp]: This function returns a result. I believe you'd want to handle `FileNotFound` and `UnknownRecordType` and return `EXIT_FAIL`.

**Review by mjtieto — CHANGES_REQUESTED** [review]

File: exercise_3/main.cpp (hunk 13 above)
Comment (anetso18) [reply exercise_3/main.cpp]: Could it be like :
LoadingResult result = loadObjectsFromFile(argv[1], exams, courses, teachers, students);

		if (result == LoadingResult::FileNotFound)
		{
			std::cerr << "File not found" << std::endl;
			return EXIT_FAILURE;
		}
		if (result == LoadingResult::UnknownRecordType)
		{
			std::cerr << "File not found" << std::endl;
			return EXIT_FAILURE;
		}

Or should I use try catch block?

Comment (mjtieto) [reply exercise_3/main.cpp]: The if/else is fine.
You could use const auto: `const auto result = loadObjectsFromFile(...);`

Comment (anetso18) [reply exercise_3/main.cpp]: Ok, thank you ;)

**Review by mjtieto — APPROVED** [review]

**Pull request merged**